### ISSUE TITLE
Multiple keyboards?

### ISSUE THREAD

**Issue by cynthia** [WICG/keyboard-map#9]
Brought up during TAG review: w3ctag/design-reviews#238 and #4

Related to #7, given that onkeyboardchange has been clarified, how will `getKeyboardMap()` work *when* we have two keyboards which are of a different layout (which I believe is possible on certain OSes), and when and how will onkeyboardchange fire?

**Comment by RByers** [comment]
Yeah, this seems like a problem to me.  The API seems predicated on the idea that there's exactly one keyboard attached.  To what extent do all desktop OSes today make a similar assumption? Related, how should this API behave on a device which typically has only on-screen keyboards (like a phone)?

One possible alternate design is to hang the keyboard map off the [InputDeviceCapabilities interface](http://wicg.github.io/InputDeviceCapabilities/#the-inputdevicecapabilities-interface).  In it's most basic form then you could get the mapping corresponding to the device that generated any keyboard input.  If we [added methods to enumerate](https://github.com/WICG/InputDeviceCapabilities/issues/31) devices, then it would be strictly more powerful than the proposal here.

I don't think we necessarily need to solve the multiple keyboard problem in V0, but perhaps the design should be one which could reasonably accommodate extension to handle it in the future?

**Comment by garykac** [comment]
I'm not aware of any OSs that can associate different layouts with different physical keyboards that are attached.

On desktops (Mac,Win,Linux), when I have a US, UK, French and Russian keyboard simultaneously connected to a machine, I can type on any of them (they all work) and the corresponding keys all act the same way (which is based on the active OS layout).

When I (with these 4 keyboard attached) switch the OS layout to US, UK, French, or Russian, then all of the keyboards will act as American, British, French or Russian as it can (given the slight difference in which keys are available).

For Android (AIUI), you have the option to specify the layout for your external keyboard, and it is applied to whatever keyboard you attach. Again, multiple connected keyboards share the same layout.

The only layout that matters for this API is the OS (software) layout and there is only one active at any time.  The OS layout is applied to whatever hardware happens to be connected.

**Comment by garykac** [comment]
WRT `onkeyboardlayoutchange`, it should fire whenever the OS layout changes. If the OS detects new hardware being plugged in and switches layouts (as I understand is the case with Android), then that layout switch will trigger the event. For OS's where that doesn't happen (desktop USB keyboards), then plugging in a new physical keyboard will not cause a layout change, nor will it trigger an event.

**Comment by garykac** [comment]
I've updated the [section in the spec relating to the `keyboardlayoutchange` event](https://wicg.github.io/keyboard-map/#keyboardlayoutchange-event) and included info about multiple physical keyboards.

Please let me know if there's more that should be covered there.

**Comment by garykac** [comment]
Update link to section about this event: https://wicg.github.io/keyboard-map/#layoutchange-event
(because it was renamed to "layoutchange")

**Comment by RByers** [comment]
Makes sense to me, thanks!

**Comment by cynthia** [comment]
Changes look good, thanks for addressing this!
